Add info about dock mode (#4566)
ArduPilot/ardupilot_wiki: common/source/docs/common-dock-mode.rst

.. _common-dock-mode:

=========
Dock Mode
=========

The DOCK mode maneuvers a vehicle towards a stationary docking target automatically.
Currently, we support docking with the help of a camera attached to a companion computer.

> The companion computer estimates the pose of the marker (ArucoMarker/AprilTag) placed at the docking station and sends

Reviewer: ```suggestion
The companion computer estimates the pose of the marker (`ArucoMarker/AprilTag <https://april.eecs.umich.edu/software/apriltag>`__  ) placed at the docking station and sends
```

`LANDING_TARGET <https://mavlink.io/en/messages/common.html#LANDING_TARGET>`__ mavlink messages to the vehicle which are then processed by the precision landing library
(`AC_PrecLand <https://github.com/ArduPilot/ardupilot/tree/master/libraries/AC_PrecLand>`__) to estimate the position of the docking target.

Successful tests have been conducted using the OpenMV camera as a companion computer to perform auto-docking.
However, any companion computer equipped with an RGB camera (like OpenMV camera, Raspberry Pi, Jetson Nano etc.) which can estimate pose

> of marker placed at docking target and send mavlink messages to the vehicle can be used.

Reviewer: ```suggestion
of marker placed at docking target and generate the  `LANDING_TARGET <https://mavlink.io/en/messages/common.html#LANDING_TARGET>`__ mavlink messages to the vehicle's autopilot can be used.
```

The OpenMV Camera
=================

OpenMV camera is a pocket sized camera board which is capable of performing various computer vision tasks.

> It can transmit mavlink messages over UART interface and hence it can be used as a companion computer to perform auto-docking.

Reviewer: ```suggestion
It can transmit the required mavlink messages over its UART interface and hence it can be used as a companion computer to perform auto-docking.
```

The camera can be purchased directly from `openmv.io <https://openmv.io/collections/cams>`__

Connection to Autopilot
=======================

> The companion computer needs to be connected to the autopilot through UART interface.

Reviewer: ```suggestion
The companion computer needs to be connected a serial port on the autopilot via its UART interface.
```

.. image:: ../../../images/openmv-pixhawk.jpg
   :target: ../_images/openmv-pixhawk.jpg
   :width: 600px

- Connect the camera’s P4, VIN and GND pins to one of the autopilot’s serial ports as shown above. In the image above the OpenMV cam is connected to a Pixhawk’s Telem2 port.

>   The connection scheme would be same for the other alternatives too. The user just needs to make a connection between

Reviewer: ```suggestion
  The connection scheme would be same for other serial ports on the autopilot. There just needs to make a connection between
```

  TX pin on the companion computer to RX pin of the telemetry port on the autopilot and power up the autopilot.
- Set :ref:`SERIAL2_PROTOCOL <SERIAL2_PROTOCOL>` = 1 (MAVLink) if using Serial2/Telem2, if connected to another serial port use the corresponding SERIALx_PROTOCOL parameter
- Set :ref:`SERIAL2_BAUD <SERIAL2_BAUD>` = 115

Generating AprilTag Markers
===========================

> .. image:: ../../../images/openmv-apriltag-generator.png

Reviewer: this image is missing from the commit

   :target: ../_images/openmv-apriltag-generator.png
   :width: 600px

The OpenMV camera has capability to detect the AprilTags. The OpenMV IDE comes with built-in AprilTag generator. The steps to generate an AprilTag using the same are:

- Go to Tools -> Machine Vision -> AprilTag generator -> TAG36H11 Family (587 Tags - Recommended). This will open a small dialog box.

> - Set the range of tags you want to generate. For eg. if you want to generate tags numbered 1 to 5, set Min to 1 and Max to 5.

Reviewer: what determines how many tags you should use?

Author: The user would generally need only one marker. However, the tag generator gives an option to generate a range of tags. 
I did reword the sentence to make it less confusing.

- Click ok and save the tags to the desired directory.

> 

Reviewer: now what? print it out? how big?and how should it be placed at the docking location?you should reference the the script and the script's structure for the tag numbering and border size....its not clear immediately from the script's comments that its of the format tagid:border length in mm

Author: Thanks for letting me know that I missed it. It is basically users decision how bigger marker he wants to use. The default as per the script is 165 mm (which is got when we simply print the generated image on an A4 sheet). However, a bigger marker can be used to increase the range.

Uploading Code to the Camera
============================

The OpenMV camera can be programmed using the OpenMV IDE. Detailed instructions to do so can be found `here <http://docs.openmv.io/openmvcam/tutorial/index.html>`__.

- Connect the OpenMV camera to your PC using  a micro USB cable.
- Open the OpenMV IDE and click the connect button at the bottom left corner of the IDE window.

> - Go to file menu, create new file and copy `this <https://github.com/shiv-tyagi/openmv-examples/blob/master/apriltag_autodock.py>`__ script to the editor window.

Reviewer: this script should be moved into Tools/scripts in the ardupilot repo and linked from there rather than your personal repo

Author: I have updated the script at openmv directly and have linked it from their repository.

Mounting to the vehicle
=======================

The camera should be mounted at the front of the vehicle. The camera board should be mounted such that the camera edge with
the lens is upwards (i.e., away from the ground). This is the default setup configuration that we think would be used
in most of the cases. However, support for mounting the camera at the back and in different mounting orientations is also provided.
If you want to mount the camera in any other configuration, please don't forget to set the related `PLND_XXX` parameters accordingly.

.. note::
    The precision landing library also relies on the vehicle's compass to construct the target vector.
    Any kind of `magnetic interference <https://ardupilot.org/rover/docs/common-magnetic-interference.html>`__ in the vehicle's compass may lead to the vehicle behaving inappropriately during the docking process.

Setup through GCS
=================

The following parameters are needed to be set to enable the precision landing using a companion computer on the rover.

-  :ref:`PLND_ENABLED <PLND_ENABLED>` = 1
-  :ref:`PLND_TYPE <PLND_TYPE>` = 1

> You would need to reboot the vehicle once you set up the above mentioned parameters.

Reviewer: need a blank line here after the bullet list

If you are mounting the camera in any configuration other than the default configuration mentioned in the previous section, please set the following parameters appropriately.

-  :ref:`PLND_CAM_POS_X <PLND_CAM_POS_X>`, :ref:`PLND_CAM_POS_Y <PLND_CAM_POS_Y>`, :ref:`PLND_CAM_POS_Z <PLND_CAM_POS_Z>` : Position of camera in body frame (in m)
-  :ref:`PLND_ORIENT <PLND_ORIENT>` : Orientation of camera on vehicles body. The camera can be mounted at the front or at the back of the vehicle's body
-  :ref:`PLND_YAW_ALIGN <PLND_YAW_ALIGN>` : Angle of rotation of camera from default configuration to mounted configuration around the axis pointing out from it's lens (in centi degrees)

The docking process can be fine tuned using the following parameters.

> - :ref:`DOCK_SPEED <DOCK_SPEED>` : The maximum speed limit at the which the vehicle can maneuver in the dock mode.

Reviewer: any ref link must have no whitespace after it and before a comma,period, or colon...lines 94-98 are generating these build errors...
